In [1]:
!pip install pandas

In [2]:
import pandas as pd

df = pd.read_csv('messy_sales.csv')

print("Shape:", df.shape)
display(df.head(10))
print(df.dtypes)

Shape: (300, 6)


,order_id,date,product,price,qty,zip
0,1254,04/06/2026,mug,11.36,4,60614
1,1057,2026-05-30,charger,14.79,3,30303
2,1150,05/06/2026,mug,5.50,1,10001
3,1066,05/30/2026,notebook,NaN,1,98101
4,1114,04/06/2026,webcam,45.59,4,90405
5,1280,04/22/2026,pen set,NaN,4,30303
6,1204,04/18/2026,charger,8.41,2,2134
7,1249,2026-05-15,keyboard,75.04,4,90405
8,1037,2026-05-19,mug,9.31,2,60614
9,1177,2026-05-13,mug,10.11,3,90210


order_id      int64
date            str
product         str
price       float64
qty           int64
zip           int64
dtype: object


Initial Observations
Missing Prices: The `price` column has missing values (`NaN`).
Date Formatting: The `date` column mixes different styles (like `MM/DD/YYYY` and `YYYY-MM-DD`).
Zip Code Lengths: Some zip codes are missing leading zeros (showing only 4 digits).

In [3]:
df.isna().sum()

order_id     0
date         0
product      0
price       12
qty          0
zip          0
dtype: int64

In [4]:

fill_value = df['price'].median()
print("Fill value used (median):", fill_value)

df['price'] = df['price'].fillna(fill_value)

print("Missing prices remaining:", df['price'].isna().sum())

Fill value used (median): 37.53
Missing prices remaining: 0


In [5]:

print("Duplicates found:", df.duplicated().sum())

df = df.drop_duplicates()

print("New shape:", df.shape)

Duplicates found: 8
New shape: (292, 6)


In [6]:

df['zip'] = df['zip'].astype(str).str.zfill(5)

print(df['zip'].head(10))

0    60614
1    30303
2    10001
3    98101
4    90405
5    30303
6    02134
7    90405
8    60614
9    90210
Name: zip, dtype: str


In [7]:

df['date'] = pd.to_datetime(df['date'], format='mixed')

print(df.dtypes)

order_id             int64
date        datetime64[us]
product                str
price              float64
qty                  int64
zip                    str
dtype: object


In [8]:

negative_qty = df[df['qty'] < 0]
display(negative_qty)
print("Count of negative quantity rows:", len(negative_qty))

,order_id,date,product,price,qty,zip
202,1140,2026-04-18,webcam,38.69,-5,98101
262,1233,2026-05-09,desk lamp,22.64,-4,10001
297,1025,2026-04-27,keyboard,47.30,-2,02116


Count of negative quantity rows: 3


In [9]:

df = df[df['qty'] >= 0]

print("Shape after removing negative quantities:", df.shape)

Shape after removing negative quantities: (289, 6)


In [10]:

df.to_csv('sales_clean.csv', index=False)

print("File successfully saved as sales_clean.csv!")

File successfully saved as sales_clean.csv!


Data Cleaning Log Summary
1. **Missing Prices**: Imputed 12 missing values in the `price` column using the median value ($37.53).
2. **Duplicates**: Removed 8 duplicate rows, reducing the total row count from 300 to 292.
3. **Zip Codes**: Converted zip codes to strings and padded them to 5 digits using leading zeros (e.g., fixing 4-digit zip codes).
4. **Dates**: Converted the `date` column into standard datetimes using mixed format parsing.
5. **Negative Quantities**: Filtered out 3 rows containing negative quantities (returns/anomalies), leaving a final cleaned shape of (289, 6).
6. **Export**: Saved the final processed dataset as `sales_clean.csv` without indices.